src/main/java/Assignment2/StudentService.java

```diff
@@ -0,0 +1,212 @@
+    // isStateChanged = 1
+    isStateChanged = true;
+    // add rollNo in Database
+    rollNoInRepo.add(model.rollNo);
+  }
+
+  void processInput(String name, String rollNo, String address, String age, String courses) {
+    // check valid
+    if (!validInput(name, rollNo, address, age, courses)) return;
+    if (rollNoInRepo.contains(Integer.parseInt(rollNo))) {
+      System.out.println("Roll no Already exist in Database");
+      return;
+    }
+    // make model object
+    StudentModel model = new StudentModel();
+    model.name = name;
+    model.rollNo = Integer.parseInt(rollNo);
+    model.address = address;
+    model.age = Integer.parseInt(age);
+    String[] SplitCourse = courses.split(" ");
+    int N = SplitCourse.length;
+    model.courses = courses.replaceAll(" ", "").toCharArray();
+    // addStudent(model)
+    addStudent(model);
+  }
+
+  void sortStudentList() {
+    // check if state changed
+    if (!isStateChanged) return;
+    // if yes then sort
+    if (currentType == sortOption.NAME) {
+      if (isAscending) studentList
+          .sort(Comparator.comparing(StudentModel::getName).thenComparing(StudentModel::getRollNo));
+      else studentList.sort(
+          Comparator.comparing(StudentModel::getName).thenComparing(StudentModel::getRollNo)
+              .reversed());
+    }
+    if (currentType == sortOption.AGE) {
+      if (isAscending) studentList.sort(Comparator.comparing(StudentModel::getAge));
+      else studentList.sort(Comparator.comparing(StudentModel::getAge).reversed());
+    }
+    if (currentType == sortOption.ROLL_NO) {
+      if (isAscending) studentList.sort(Comparator.comparing(StudentModel::getRollNo));
+      else studentList.sort(Comparator.comparing(StudentModel::getRollNo).reversed());
+    }
+    isStateChanged = false;
+  }
+
+  void sortPreference(sortOption sortType, boolean isAscending) {
+    // same then nothing
+    if (sortType == currentType && isAscending == this.isAscending) return;
+    currentType = sortType;
+    this.isAscending = isAscending;
+    isStateChanged = true;
+  }
+
+  void showData() {
+    // sort before display
+    sortStudentList();
+    // print data in studentList
+    for (StudentModel student : studentList) {
+      student.print();
+    }
+  }
+
+  void deleteStudent(String rollNo) {
+    // validating input
+    if (rollNo.isEmpty() || !rollNo.matches("[0-9]*")) {
+      System.out.println("Invalid quantity");
+      return;
+    }
+    // checking if rollNo present in studentList or not
+    if (!rollNoInRepo.contains(Integer.parseInt(rollNo))) {
+      System.out.println("RollNo not present");
+      return;
+    }
+    // delete data from studentList
+    for (int i = 0; i < studentList.size(); i++) {
+      if (studentList.get(i).rollNo == Integer.parseInt(rollNo)) {
+        studentList.remove(i);
+        rollNoInRepo.remove(Integer.parseInt(rollNo));
+        isStateChanged = true;
+        break;
+      }
+    }
+  }
+
+  void saveState() {
```

Comment:
every call of saveState() and getState() is creating repo object means db object. prevent this duplicate object creation.